# Fetch Rewards Coding Exercise - Data Analyst




###### The ER Diagram is attached as a separate file.
###### The data is converted from JSON 

In [1]:
# importing packages
import json
import pandas as pd
from pandasql import sqldf
import pandasql as ps

In [177]:
# Converting Brands data from json file
df1 = open(r'brands.json')
brands = [json.loads(line) for line in df1]

# Converting users data from json file
df2 = open(r'users.json')
users = [json.loads(line) for line in df2]
    
# Converting receipts data from json file    
df3 = open(r'receipts.json')
receipts = [json.loads(line) for line in df3]

In [178]:
# Normalizaing the data
df_brands = pd.json_normalize(brands)
df_users = pd.json_normalize(users)
df_receipts = pd.json_normalize(receipts)

In [5]:
# It prints out the first five rows of the dataset
df_brands.head()

barcode        category      categoryCode                       name  \
0  511111019862          Baking            BAKING  test brand @1612366101024   
1  511111519928       Beverages         BEVERAGES                  Starbucks   
2  511111819905          Baking            BAKING  test brand @1612366146176   
3  511111519874          Baking            BAKING  test brand @1612366146051   
4  511111319917  Candy & Sweets  CANDY_AND_SWEETS  test brand @1612366146827   

  topBrand                  _id.$oid              cpg.$id.$oid cpg.$ref  \
0    False  601ac115be37ce2ead437551  601ac114be37ce2ead437550     Cogs   
1    False  601c5460be37ce2ead43755f  5332f5fbe4b03c9a25efd0ba     Cogs   
2    False  601ac142be37ce2ead43755d  601ac142be37ce2ead437559     Cogs   
3    False  601ac142be37ce2ead43755a  601ac142be37ce2ead437559     Cogs   
4    False  601ac142be37ce2ead43755e  5332fa12e4b03c9a25efd1e7     Cogs   

                       brandCode  
0                            NaN  
1                      STARBUCKS  
2  TEST BRANDCODE @1612366146176  
3  TEST BRANDCODE @1612366146051  
4  TEST BRANDCODE @1612366146827

In [6]:
df_users.head()

active      role signUpSource state                  _id.$oid  \
0    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
1    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
2    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
3    True  consumer        Email    WI  5ff1e1eacfcf6c399c274ae6   
4    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   

   createdDate.$date  lastLogin.$date  
0      1609687444800     1.609688e+12  
1      1609687444800     1.609688e+12  
2      1609687444800     1.609688e+12  
3      1609687530554     1.609688e+12  
4      1609687444800     1.609688e+12

In [7]:
df_receipts.head()

bonusPointsEarned                            bonusPointsEarnedReason  \
0              500.0  Receipt number 2 completed, bonus point schedu...   
1              150.0  Receipt number 5 completed, bonus point schedu...   
2                5.0                         All-receipts receipt bonus   
3                5.0                         All-receipts receipt bonus   
4                5.0                         All-receipts receipt bonus   

  pointsEarned  purchasedItemCount  \
0        500.0                 5.0   
1        150.0                 2.0   
2            5                 1.0   
3          5.0                 4.0   
4          5.0                 2.0   

                              rewardsReceiptItemList rewardsReceiptStatus  \
0  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
1  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
2  [{'needsFetchReview': False, 'partnerItemId': ...             REJECTED   
3  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
4  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   

  totalSpent                    userId                  _id.$oid  \
0      26.00  5ff1e1eacfcf6c399c274ae6  5ff1e1eb0a720f0523000575   
1      11.00  5ff1e194b6a9d73a3a9f1052  5ff1e1bb0a720f052300056b   
2      10.00  5ff1e1f1cfcf6c399c274b0b  5ff1e1f10a720f052300057a   
3      28.00  5ff1e1eacfcf6c399c274ae6  5ff1e1ee0a7214ada100056f   
4       1.00  5ff1e194b6a9d73a3a9f1052  5ff1e1d20a7214ada1000561   

   createDate.$date  dateScanned.$date  finishedDate.$date  modifyDate.$date  \
0     1609687531000      1609687531000        1.609688e+12     1609687536000   
1     1609687483000      1609687483000        1.609687e+12     1609687488000   
2     1609687537000      1609687537000                 NaN     1609687542000   
3     1609687534000      1609687534000        1.609688e+12     1609687539000   
4     1609687506000      1609687506000        1.609688e+12     1609687511000   

   pointsAwardedDate.$date  purchaseDate.$date  
0             1.609688e+12        1.609632e+12  
1             1.609687e+12        1.609601e+12  
2                      NaN        1.609632e+12  
3             1.609688e+12        1.609632e+12  
4             1.609688e+12        1.609601e+12

In [8]:
#Renaming the column names for easy identification
df_brands = df_brands.rename(columns={'_id.$oid':'id','cpg.$id.$oid':'cpgid','cpg.$ref':'cpgref'})

# Converting the Users data which is in format 'epoch' to 'datetime' and renaming the columns
df_users['createdDate.$date'] = pd.to_datetime(df_users['createdDate.$date']).dt.date
df_users['lastLogin.$date'] = pd.to_datetime(df_users['lastLogin.$date']).dt.date
df_users = df_users.rename(columns={'_id.$oid':'id','createdDate.$date':'createdDate','lastLogin.$date': 'lastLogin'})

# Converting the Receipts data which is in format 'epoch' to 'datetime'
df_receipts['createdDate.$date'] = pd.to_datetime(df_receipts['createDate.$date']).dt.date
df_receipts['dateScanned.$date'] = pd.to_datetime(df_receipts['dateScanned.$date']).dt.date
df_receipts['finishedDate.$date'] = pd.to_datetime(df_receipts['finishedDate.$date']).dt.date
df_receipts['modifyDate.$date'] = pd.to_datetime(df_receipts['modifyDate.$date']).dt.date
df_receipts['pointsAwardedDate.$date'] = pd.to_datetime(df_receipts['pointsAwardedDate.$date']).dt.date
df_receipts['purchaseDate.$date'] = pd.to_datetime(df_receipts['purchaseDate.$date']).dt.date


In [9]:
# Renamed some columns in receipts to make it easier
df_receipts = df_receipts.rename(columns={'_id.$oid':'id','createDate.$date':'createdDate','dateScanned.$date':'dateScanned','finishedDate.$date':'finishedDate','modifyDate.$date':'modifyDate','pointsAwardedDate.$date':'pointsAwardedDate','purchaseDate.$date':'purchaseDate'})
df_receipts.head()

bonusPointsEarned                            bonusPointsEarnedReason  \
0              500.0  Receipt number 2 completed, bonus point schedu...   
1              150.0  Receipt number 5 completed, bonus point schedu...   
2                5.0                         All-receipts receipt bonus   
3                5.0                         All-receipts receipt bonus   
4                5.0                         All-receipts receipt bonus   

  pointsEarned  purchasedItemCount  \
0        500.0                 5.0   
1        150.0                 2.0   
2            5                 1.0   
3          5.0                 4.0   
4          5.0                 2.0   

                              rewardsReceiptItemList rewardsReceiptStatus  \
0  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
1  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
2  [{'needsFetchReview': False, 'partnerItemId': ...             REJECTED   
3  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
4  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   

  totalSpent                    userId                        id  \
0      26.00  5ff1e1eacfcf6c399c274ae6  5ff1e1eb0a720f0523000575   
1      11.00  5ff1e194b6a9d73a3a9f1052  5ff1e1bb0a720f052300056b   
2      10.00  5ff1e1f1cfcf6c399c274b0b  5ff1e1f10a720f052300057a   
3      28.00  5ff1e1eacfcf6c399c274ae6  5ff1e1ee0a7214ada100056f   
4       1.00  5ff1e194b6a9d73a3a9f1052  5ff1e1d20a7214ada1000561   

     createdDate dateScanned finishedDate  modifyDate pointsAwardedDate  \
0  1609687531000  1970-01-01   1970-01-01  1970-01-01        1970-01-01   
1  1609687483000  1970-01-01   1970-01-01  1970-01-01        1970-01-01   
2  1609687537000  1970-01-01          NaT  1970-01-01               NaT   
3  1609687534000  1970-01-01   1970-01-01  1970-01-01        1970-01-01   
4  1609687506000  1970-01-01   1970-01-01  1970-01-01        1970-01-01   

  purchaseDate createdDate.$date  
0   1970-01-01        1970-01-01  
1   1970-01-01        1970-01-01  
2   1970-01-01        1970-01-01  
3   1970-01-01        1970-01-01  
4   1970-01-01        1970-01-01

## Data Exploration/ Data Profiling
#### Data profiling helps in knowing the structure of the data, discovering content in order to perform data pre-processing for further analysis and producing visulizations.

#### Here I am using some basic commands to understand the dataset and attributes in the dataframe

In [10]:
df_brands.shape

(1167, 9)

In [11]:
#Brands Dataframe 
print("Number of Rows in the dataframe are:", len(df_brands))
print("Number of columns in the dataframe are:", len(df_brands.columns))

Number of Rows in the dataframe are: 1167
Number of columns in the dataframe are: 9


In [12]:
df_users.shape

(495, 7)

In [13]:
#UsersDataframe 
print("Number of Rows in the dataframe are:", len(df_users))
print("Number of columns in the dataframe are:", len(df_users.columns))

Number of Rows in the dataframe are: 495
Number of columns in the dataframe are: 7


In [14]:
df_receipts.shape

(1119, 16)

In [15]:
#Receipts Dataframe 
print("Number of Rows in the dataframe are:", len(df_receipts))
print("Number of columns in the dataframe are:", len(df_receipts.columns))

Number of Rows in the dataframe are: 1119
Number of columns in the dataframe are: 16


In [16]:
# Count of each variable
df_brands.count()

barcode         1167
category        1012
categoryCode     517
name            1167
topBrand         555
id              1167
cpgid           1167
cpgref          1167
brandCode        933
dtype: int64

In [17]:
df_brands.describe()

barcode category categoryCode       name topBrand  \
count           1167     1012          517       1167      555   
unique          1160       23           14       1156        2   
top     511111204923   Baking       BAKING  Baken-Ets    False   
freq               2      369          359          2      524   

                              id                     cpgid cpgref brandCode  
count                       1167                      1167   1167       933  
unique                      1167                       196      2       897  
top     5da605bea60b87376833e34a  559c2234e4b06aca36af13c6   Cogs            
freq                           1                        98   1020        35

In [18]:
# These are the list of unique categories which are there in the 'brands' dataframe.
df_brands['category'].unique()

array(['Baking', 'Beverages', 'Candy & Sweets', 'Condiments & Sauces',
       'Canned Goods & Soups', nan, 'Magazines', 'Breakfast & Cereal',
       'Beer Wine Spirits', 'Health & Wellness', 'Beauty', 'Baby',
       'Frozen', 'Grocery', 'Snacks', 'Household', 'Personal Care',
       'Dairy', 'Cleaning & Home Improvement', 'Deli',
       'Beauty & Personal Care', 'Bread & Bakery', 'Outdoor',
       'Dairy & Refrigerated'], dtype=object)

In [19]:
# There are nan values here, which might affect the analysis to see which is the topBrand.
df_brands['topBrand'].unique()

array([False, nan, True], dtype=object)

In [20]:
df_brands['barcode'].duplicated().sum()

7

### We can observe that there are 7 barcodes which are duplicated
#### The top category is Baking and there are 23 different categories.
##### Caleb's Kola is the top brand 

In [21]:
df_users.count()

active          495
role            495
signUpSource    447
state           439
id              495
createdDate     495
lastLogin       433
dtype: int64

In [22]:
df_users.describe()

active      role signUpSource state                        id  \
count     495       495          447   439                       495   
unique      2         2            2     8                       212   
top      True  consumer        Email    WI  54943462e4b07e684157a532   
freq      494       413          443   396                        20   

       createdDate   lastLogin  
count          495         433  
unique           1           1  
top     1970-01-01  1970-01-01  
freq           495         433

##### Here, we can see that there are two unique roles, upon which consumer is top.
#### Email is the top source for signUp.

###### WI has the highest Signup's for Fetch Rewards.

###### There are a lot of nan values, which sum's up almost to 50% of the data being nan's, this will not produce accurate analysis. Let's say we take states. The dataset shows that there are signup's from only 8 states, but in reality those nan's might be other states. 

In [23]:
# There is no missing data for role.
df_users['role'].unique()

array(['consumer', 'fetch-staff'], dtype=object)

In [24]:
df_users['signUpSource'].unique()

array(['Email', 'Google', nan], dtype=object)

In [25]:
df_users['state'].unique()

array(['WI', 'KY', 'AL', 'CO', 'IL', nan, 'OH', 'SC', 'NH'], dtype=object)

In [26]:
df_receipts.count()

bonusPointsEarned           544
bonusPointsEarnedReason     544
pointsEarned                609
purchasedItemCount          635
rewardsReceiptItemList      679
rewardsReceiptStatus       1119
totalSpent                  684
userId                     1119
id                         1119
createdDate                1119
dateScanned                1119
finishedDate                568
modifyDate                 1119
pointsAwardedDate           537
purchaseDate                671
createdDate.$date          1119
dtype: int64

In [27]:
df_receipts.describe()

bonusPointsEarned  purchasedItemCount   createdDate
count         544.000000           635.00000  1.119000e+03
mean          238.893382            14.75748  1.611800e+12
std           299.091731            61.13424  1.484091e+09
min             5.000000             0.00000  1.604089e+12
25%             5.000000             1.00000  1.610652e+12
50%            45.000000             2.00000  1.611941e+12
75%           500.000000             5.00000  1.612704e+12
max           750.000000           689.00000  1.614641e+12

In [98]:
df_receipts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1119 entries, 0 to 1118
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   bonusPointsEarned        544 non-null    float64
 1   bonusPointsEarnedReason  544 non-null    object 
 2   pointsEarned             609 non-null    object 
 3   purchasedItemCount       635 non-null    float64
 4   rewardsReceiptItemList   679 non-null    object 
 5   rewardsReceiptStatus     1119 non-null   object 
 6   totalSpent               684 non-null    object 
 7   userId                   1119 non-null   object 
 8   id                       1119 non-null   object 
 9   createdDate              1119 non-null   int64  
 10  dateScanned              1119 non-null   object 
 11  finishedDate             568 non-null    object 
 12  modifyDate               1119 non-null   object 
 13  pointsAwardedDate        537 non-null    object 
 14  purchaseDate            

In [99]:
df_receipts['id'].nunique()

1119

In [100]:
df_users.head()

active      role signUpSource state                        id createdDate  \
0    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052  1970-01-01   
1    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052  1970-01-01   
2    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052  1970-01-01   
3    True  consumer        Email    WI  5ff1e1eacfcf6c399c274ae6  1970-01-01   
4    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052  1970-01-01   

    lastLogin  
0  1970-01-01  
1  1970-01-01  
2  1970-01-01  
3  1970-01-01  
4  1970-01-01

In [101]:
# TO Find missing values in brands Dataframe
df_brands.isnull().sum()

barcode           0
category        155
categoryCode    650
name              0
topBrand        612
id                0
cpgid             0
cpgref            0
brandCode       234
dtype: int64

In [102]:
# TO Find missing values in users Dataframe
df_users.isnull().sum()

active           0
role             0
signUpSource    48
state           56
id               0
createdDate      0
lastLogin       62
dtype: int64

In [103]:
# TO Find missing values in receipts Dataframe
df_receipts.isnull().sum()

bonusPointsEarned          575
bonusPointsEarnedReason    575
pointsEarned               510
purchasedItemCount         484
rewardsReceiptItemList     440
rewardsReceiptStatus         0
totalSpent                 435
userId                       0
id                           0
createdDate                  0
dateScanned                  0
finishedDate               551
modifyDate                   0
pointsAwardedDate          582
purchaseDate               448
createdDate.$date            0
dtype: int64

In [104]:
df_brands.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1167 entries, 0 to 1166
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   barcode       1167 non-null   object
 1   category      1012 non-null   object
 2   categoryCode  517 non-null    object
 3   name          1167 non-null   object
 4   topBrand      555 non-null    object
 5   id            1167 non-null   object
 6   cpgid         1167 non-null   object
 7   cpgref        1167 non-null   object
 8   brandCode     933 non-null    object
dtypes: object(9)
memory usage: 82.2+ KB


In [105]:
df_receipts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1119 entries, 0 to 1118
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   bonusPointsEarned        544 non-null    float64
 1   bonusPointsEarnedReason  544 non-null    object 
 2   pointsEarned             609 non-null    object 
 3   purchasedItemCount       635 non-null    float64
 4   rewardsReceiptItemList   679 non-null    object 
 5   rewardsReceiptStatus     1119 non-null   object 
 6   totalSpent               684 non-null    object 
 7   userId                   1119 non-null   object 
 8   id                       1119 non-null   object 
 9   createdDate              1119 non-null   int64  
 10  dateScanned              1119 non-null   object 
 11  finishedDate             568 non-null    object 
 12  modifyDate               1119 non-null   object 
 13  pointsAwardedDate        537 non-null    object 
 14  purchaseDate            

In [106]:
df_users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   active        495 non-null    bool  
 1   role          495 non-null    object
 2   signUpSource  447 non-null    object
 3   state         439 non-null    object
 4   id            495 non-null    object
 5   createdDate   495 non-null    object
 6   lastLogin     433 non-null    object
dtypes: bool(1), object(6)
memory usage: 23.8+ KB


In [107]:
# To find duplicates in brands, users and receipts table.
print(df_users.duplicated().sum()) 
print(df_brands.duplicated().sum())
print(df_receipts.duplicated(subset = ['id']).sum())

283
0
0


In [108]:
print(df_users['id'].duplicated().sum()) 

283


### Out of 495 users, 283 are duplicates.

In [109]:
df_users.nunique()

active            2
role              2
signUpSource      2
state             8
id              212
createdDate       1
lastLogin         1
dtype: int64

In [110]:
uniqueids = df_brands['id'].nunique()
print('Number of unique ids in column "id" in users dataframe :', uniqueids)

Number of unique ids in column "id" in users dataframe : 1167


In [111]:
# 7 barcodes are duplicate values
df_brands.nunique()

barcode         1160
category          23
categoryCode      14
name            1156
topBrand           2
id              1167
cpgid            196
cpgref             2
brandCode        897
dtype: int64

In [112]:
# Unique Values in the Brands dataframe
uniquebarcodes = df_brands['barcode'].nunique()
uniquecat = df_brands['category'].nunique()
uniquename = df_brands['name'].nunique()
uniqueid = df_brands['id'].nunique()
uniquebrandcode = df_brands['brandCode'].nunique()
print('Number of unique values in column "barcode" in brands dataframe :', uniquebarcodes)
print('Number of unique values in column "category" in brands dataframe :', uniquecat)
print('Number of unique values in column "name" in brands dataframe :', uniquename)
print('Number of unique values in column "id" in brands dataframe :', uniqueid)
print('Number of unique values in column "brandCode" in brands dataframe :', uniquebrandcode)

Number of unique values in column "barcode" in brands dataframe : 1160
Number of unique values in column "category" in brands dataframe : 23
Number of unique values in column "name" in brands dataframe : 1156
Number of unique values in column "id" in brands dataframe : 1167
Number of unique values in column "brandCode" in brands dataframe : 897


In [113]:
df_receipts

bonusPointsEarned                            bonusPointsEarnedReason  \
0                 500.0  Receipt number 2 completed, bonus point schedu...   
1                 150.0  Receipt number 5 completed, bonus point schedu...   
2                   5.0                         All-receipts receipt bonus   
3                   5.0                         All-receipts receipt bonus   
4                   5.0                         All-receipts receipt bonus   
...                 ...                                                ...   
1114               25.0                        COMPLETE_NONPARTNER_RECEIPT   
1115                NaN                                                NaN   
1116                NaN                                                NaN   
1117               25.0                        COMPLETE_NONPARTNER_RECEIPT   
1118                NaN                                                NaN   

     pointsEarned  purchasedItemCount  \
0           500.0                 5.0   
1           150.0                 2.0   
2               5                 1.0   
3             5.0                 4.0   
4             5.0                 2.0   
...           ...                 ...   
1114         25.0                 2.0   
1115          NaN                 NaN   
1116          NaN                 NaN   
1117         25.0                 2.0   
1118          NaN                 NaN   

                                 rewardsReceiptItemList rewardsReceiptStatus  \
0     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
1     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
2     [{'needsFetchReview': False, 'partnerItemId': ...             REJECTED   
3     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
4     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
...                                                 ...                  ...   
1114  [{'barcode': 'B076FJ92M4', 'description': 'mue...             REJECTED   
1115                                                NaN            SUBMITTED   
1116                                                NaN            SUBMITTED   
1117  [{'barcode': 'B076FJ92M4', 'description': 'mue...             REJECTED   
1118                                                NaN            SUBMITTED   

     totalSpent                    userId                        id  \
0         26.00  5ff1e1eacfcf6c399c274ae6  5ff1e1eb0a720f0523000575   
1         11.00  5ff1e194b6a9d73a3a9f1052  5ff1e1bb0a720f052300056b   
2         10.00  5ff1e1f1cfcf6c399c274b0b  5ff1e1f10a720f052300057a   
3         28.00  5ff1e1eacfcf6c399c274ae6  5ff1e1ee0a7214ada100056f   
4          1.00  5ff1e194b6a9d73a3a9f1052  5ff1e1d20a7214ada1000561   
...         ...                       ...                       ...   
1114      34.96  5fc961c3b8cfca11a077dd33  603cc0630a720fde100003e6   
1115        NaN  5fc961c3b8cfca11a077dd33  603d0b710a720fde1000042a   
1116        NaN  5fc961c3b8cfca11a077dd33  603cf5290a720fde10000413   
1117      34.96  5fc961c3b8cfca11a077dd33  603ce7100a7217c72c000405   
1118        NaN  5fc961c3b8cfca11a077dd33  603c4fea0a7217c72c000389   

        createdDate dateScanned finishedDate  modifyDate pointsAwardedDate  \
0     1609687531000  1970-01-01   1970-01-01  1970-01-01        1970-01-01   
1     1609687483000  1970-01-01   1970-01-01  1970-01-01        1970-01-01   
2     1609687537000  1970-01-01          NaT  1970-01-01               NaT   
3     1609687534000  1970-01-01   1970-01-01  1970-01-01        1970-01-01   
4     1609687506000  1970-01-01   1970-01-01  1970-01-01        1970-01-01   
...             ...         ...          ...         ...               ...   
1114  1614594147000  1970-01-01          NaT  1970-01-01               NaT   
1115  1614613361873  1970-01-01          NaT  1970-01-01               NaT   
1116  1614607657664  1970-01-01          NaT  1970-01-01    

#### "rewardsReceiptItemList" column had nested data in it.
##### created a separate new dataframe to check the data.

In [114]:
#Seperating "rewardsReceiptItemList" column and converting it into new dataFrame to check the data Quality issues
[elem.update({'rewardsReceiptItemList':[]}) for elem in receipts if 'rewardsReceiptItemList' not in elem.keys()]
rewardsReceiptItemList = pd.json_normalize(data=receipts, record_path=['rewardsReceiptItemList'], 
                              meta = [['_id', '$oid']], errors ='ignore')
rewardsReceiptItemList

barcode                                        description  \
0             4011                                     ITEM NOT FOUND   
1             4011                                     ITEM NOT FOUND   
2     028400642255  DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...   
3              NaN                                                NaN   
4             4011                                     ITEM NOT FOUND   
...            ...                                                ...   
6936    B07BRRLSVC  thindust summer face mask - sun protection nec...   
6937    B076FJ92M4  mueller austria hypergrind precision electric ...   
6938    B07BRRLSVC  thindust summer face mask - sun protection nec...   
6939    B076FJ92M4  mueller austria hypergrind precision electric ...   
6940    B07BRRLSVC  thindust summer face mask - sun protection nec...   

     finalPrice itemPrice needsFetchReview partnerItemId  \
0         26.00     26.00            False             1   
1             1         1              NaN             1   
2         10.00     10.00             True             2   
3           NaN       NaN            False             1   
4         28.00     28.00            False             1   
...         ...       ...              ...           ...   
6936      11.99     11.99              NaN             1   
6937      22.97     22.97              NaN             0   
6938      11.99     11.99              NaN             1   
6939      22.97     22.97              NaN             0   
6940      11.99     11.99              NaN             1   

     preventTargetGapPoints  quantityPurchased userFlaggedBarcode  \
0                      True                5.0               4011   
1                       NaN                1.0                NaN   
2                      True                1.0       028400642255   
3                      True                NaN               4011   
4                      True                4.0               4011   
...                     ...                ...                ...   
6936                    NaN                1.0                NaN   
6937                    NaN                1.0                NaN   
6938                    NaN                1.0                NaN   
6939                    NaN                1.0                NaN   
6940                    NaN                1.0                NaN   

     userFlaggedNewItem  ... originalMetaBriteQuantityPurchased  pointsEarned  \
0                  True  ...                                NaN           NaN   
1                   NaN  ...                                NaN           NaN   
2                  True  ...                                NaN           NaN   
3                  True  ...                                NaN           NaN   
4                  True  ...                                NaN           NaN   
...                 ...  ...                                ...           ...   
6936                NaN  ...                                NaN           NaN   
6937                NaN  ...                                NaN           NaN   
6938                NaN  ...                                NaN           NaN   
6939                NaN  ...                                NaN           NaN   
6940                NaN  ...                                NaN           NaN   

     targetPrice competitiveProduct originalFinalPrice  \
0            NaN                NaN                NaN   
1            NaN                NaN                NaN   
2            NaN                NaN                NaN   
3            NaN                NaN                NaN   
4            NaN                NaN                NaN   
...          ...                ...                ...   
6936         NaN                NaN                NaN   
6937         NaN                NaN                NaN   
6938         NaN                NaN                NaN   
6939         NaN               

### Number of Rows and Columns in the dataframe 

In [115]:
rewardsReceiptItemList.shape

(6941, 35)

### Number of Null Values in Each Column

In [116]:
rewardsReceiptItemList.isnull().sum()

barcode                               3851
description                            381
finalPrice                             174
itemPrice                              174
needsFetchReview                      6128
partnerItemId                            0
preventTargetGapPoints                6583
quantityPurchased                      174
userFlaggedBarcode                    6604
userFlaggedNewItem                    6618
userFlaggedPrice                      6642
userFlaggedQuantity                   6642
needsFetchReviewReason                6722
pointsNotAwardedReason                6601
pointsPayerId                         5674
rewardsGroup                          5210
rewardsProductPartnerId               4672
userFlaggedDescription                6736
originalMetaBriteBarcode              6870
originalMetaBriteDescription          6931
brandCode                             4341
competitorRewardsGroup                6666
discountedItemPrice                   1172
originalRec

### We can observe that there is more than 80% of the data missing from all the 35 columns present in the "rewardsReceiptItemList" and needs major data re-modleing and removing all the unwanted columns in the list. 
### Having such high amount of missing data can cause major issues in the futher data precessing and Analysis.

#### Created a dict to store the itemlist against its receiptId

In [117]:
receiptitems = {}
for index , receipt in df_receipts.iterrows():
    #check for nan
    if receipt['rewardsReceiptItemList'] != receipt['rewardsReceiptItemList']:
        pass
    else:
        receiptitems[receipt['id']] = receipt['rewardsReceiptItemList']

In [120]:
#create a pandas dataframe to flatten the receiptsRewardItemList and add a receiptId column with it
current_row = {}
list_df = pd.DataFrame()

for key in receiptitems:
    for val in receiptitems[key]:
        current_row['receiptId'] = key
        for i in val:
            current_row[i] = val[i]
    list_df = list_df.append(current_row , ignore_index = True)
list_df.head()

barcode                                        description finalPrice  \
0          4011                                     ITEM NOT FOUND      26.00   
1  028400642255  DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...      10.00   
2  028400642255  DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...      10.00   
3          4011                                     ITEM NOT FOUND      28.00   
4          1234                                     ITEM NOT FOUND       2.56   

  itemPrice  needsFetchReview partnerItemId  preventTargetGapPoints  \
0     26.00               0.0             1                     1.0   
1     10.00               1.0             2                     1.0   
2     10.00               0.0             1                     1.0   
3     28.00               0.0             1                     1.0   
4      2.56               1.0             2                     1.0   

   quantityPurchased                 receiptId userFlaggedBarcode  ...  \
0                5.0  5ff1e1eb0a720f0523000575               4011  ...   
1                1.0  5ff1e1bb0a720f052300056b       028400642255  ...   
2                1.0  5ff1e1f10a720f052300057a               4011  ...   
3                4.0  5ff1e1ee0a7214ada100056f               4011  ...   
4                3.0  5ff1e1d20a7214ada1000561               1234  ...   

   itemNumber originalMetaBriteQuantityPurchased  pointsEarned targetPrice  \
0         NaN                                NaN           NaN         NaN   
1         NaN                                NaN           NaN         NaN   
2         NaN                                NaN           NaN         NaN   
3         NaN                                NaN           NaN         NaN   
4         NaN                                NaN           NaN         NaN   

  competitiveProduct originalFinalPrice originalMetaBriteItemPrice deleted  \
0                NaN                NaN                        NaN     NaN   
1                NaN                NaN                        NaN     NaN   
2                NaN                NaN                        NaN     NaN   
3                NaN                NaN                        NaN     NaN   
4                NaN                NaN                        NaN     NaN   

  priceAfterCoupon metabriteCampaignId  
0              NaN                 NaN  
1              NaN                 NaN  
2              NaN                 NaN  
3              NaN                 NaN  
4              NaN                 NaN  

[5 rows x 35 columns]

In [121]:
list_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 679 entries, 0 to 678
Data columns (total 35 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   barcode                             679 non-null    object 
 1   description                         679 non-null    object 
 2   finalPrice                          679 non-null    object 
 3   itemPrice                           679 non-null    object 
 4   needsFetchReview                    679 non-null    float64
 5   partnerItemId                       679 non-null    object 
 6   preventTargetGapPoints              679 non-null    float64
 7   quantityPurchased                   679 non-null    float64
 8   receiptId                           679 non-null    object 
 9   userFlaggedBarcode                  679 non-null    object 
 10  userFlaggedNewItem                  679 non-null    float64
 11  userFlaggedPrice                    679 non-n

In [122]:
#Count of Items scanned for a particular barcode.
list_df['brandCode'].value_counts()

BRAND                        196
MISSION                      121
KRAFT EASY CHEESE             59
WINGSTOP                      47
BEN AND JERRYS                38
GERM-X                        33
AMERICAN BEAUTY               27
BORDEN                        24
KLEENEX                       24
CALIFIA FARMS                 15
KLARBRUNN                     14
KRUSTEAZ                      12
VIVA                           9
HUGGIES                        7
GALLO FAMILY VINEYARDS         7
KIT KAT                        7
OSCAR MAYER                    7
FOLGERS                        4
KNORR                          3
DORITOS                        2
MCCORMICK GRILL MATES          1
STOUFFER'S CLASSICS            1
WONDERFUL                      1
KARO                           1
THOMAS ENGLISH MUFFINS         1
CHEERIOS                       1
PEPSI                          1
SKITTLES                       1
DOLE CHILLED FRUIT JUICES      1
THAI KITCHEN                   1
HY-VEE    

In [123]:
df_receipts['rewardsReceiptStatus'].value_counts()

FINISHED     518
SUBMITTED    434
REJECTED      71
PENDING       50
FLAGGED       46
Name: rewardsReceiptStatus, dtype: int64

In [124]:
df_brands

barcode            category      categoryCode  \
0     511111019862              Baking            BAKING   
1     511111519928           Beverages         BEVERAGES   
2     511111819905              Baking            BAKING   
3     511111519874              Baking            BAKING   
4     511111319917      Candy & Sweets  CANDY_AND_SWEETS   
...            ...                 ...               ...   
1162  511111116752              Baking            BAKING   
1163  511111706328  Breakfast & Cereal               NaN   
1164  511111416173      Candy & Sweets  CANDY_AND_SWEETS   
1165  511111400608             Grocery               NaN   
1166  511111019930              Baking            BAKING   

                           name topBrand                        id  \
0     test brand @1612366101024    False  601ac115be37ce2ead437551   
1                     Starbucks    False  601c5460be37ce2ead43755f   
2     test brand @1612366146176    False  601ac142be37ce2ead43755d   
3     test brand @1612366146051    False  601ac142be37ce2ead43755a   
4     test brand @1612366146827    False  601ac142be37ce2ead43755e   
...                         ...      ...                       ...   
1162  test brand @1601644365844      NaN  5f77274dbe37ce6b592e90c0   
1163        Dippin Dots® Cereal      NaN  5dc1fca91dda2c0ad7da64ae   
1164  test brand @1598639215217      NaN  5f494c6e04db711dd8fe87e7   
1165          LIPTON TEA Leaves    False  5a021611e4b00efe02b02a57   
1166  test brand @1613158231643    False  6026d757be37ce6369301468   

                         cpgid cpgref                      brandCode  
0     601ac114be37ce2ead437550   Cogs                            NaN  
1     5332f5fbe4b03c9a25efd0ba   Cogs                      STARBUCKS  
2     601ac142be37ce2ead437559   Cogs  TEST BRANDCODE @1612366146176  
3     601ac142be37ce2ead437559   Cogs  TEST BRANDCODE @1612366146051  
4     5332fa12e4b03c9a25efd1e7   Cogs  TEST BRANDCODE @1612366146827  
...                        ...    ...                            ...  
1162  5f77274dbe37ce6b592e90bf   Cogs                            NaN  
1163  53e10d6368abd3c7065097cc   Cogs             DIPPIN DOTS CEREAL  
1164  5332fa12e4b03c9a25efd1e7   Cogs  TEST BRANDCODE @1598639215217  
1165  5332f5f6e4b03c9a25efd0b4   Cogs              LIPTON TEA Leaves  
1166  6026d757be37ce6369301467   Cogs  TEST BRANDCODE @1613158231644  

[1167 rows x 9 columns]

In [125]:
df_brands.isnull().sum()

barcode           0
category        155
categoryCode    650
name              0
topBrand        612
id                0
cpgid             0
cpgref            0
brandCode       234
dtype: int64

In [143]:
#We have to treat the null values in the dataset
df_brands_1 = df_brands.dropna()

##### By dropping all the rows with null values; more than 80% of data is being removed which is not feasible for analysis.

In [127]:
# By dropping rows with null values more than 80% of data is being removed which is not feasible for analysis.
df_brands_1['name'].describe()

count                           255
unique                          255
top       test brand @1609794603236
freq                              1
Name: name, dtype: object

In [149]:
df_brands.columns

Index(['barcode', 'category', 'categoryCode', 'name', 'topBrand', 'id',
       'cpgid', 'cpgref', 'brandCode'],
      dtype='object')

In [181]:
df_brands['name'].describe()

count          1167
unique         1156
top       Baken-Ets
freq              2
Name: name, dtype: object

In [231]:
df_brands['name'].nunique()

1156

In [230]:
display(df_brands['name'].iloc[582])

'Gerber Good Start'

In [232]:
df_brands['topBrand'].unique()

array([False, nan, True], dtype=object)

In [239]:
df_brands_topBrand = df_brands['topBrand'],df_brands['name']

In [237]:
df_brands_topBrand

(0       False
 1       False
 2       False
 3       False
 4       False
         ...  
 1162      NaN
 1163      NaN
 1164      NaN
 1165    False
 1166    False
 Name: topBrand, Length: 1167, dtype: object,
 0       test brand @1612366101024
 1                       Starbucks
 2       test brand @1612366146176
 3       test brand @1612366146051
 4       test brand @1612366146827
                   ...            
 1162    test brand @1601644365844
 1163          Dippin Dots® Cereal
 1164    test brand @1598639215217
 1165            LIPTON TEA Leaves
 1166    test brand @1613158231643
 Name: name, Length: 1167, dtype: object)

##### Category Code and topBrand are removed as they contain more than 70% of null data and these columns cannot be used for our analysis.

In [246]:
# Set the limit
# Drop columns using that limit
limit = len(df_brands) * 0.7
new_brands = df_brands.dropna(axis=1, thresh=limit)

In [141]:
new_brands.columns

Index(['barcode', 'category', 'name', 'id', 'cpgid', 'cpgref', 'brandCode'], dtype='object')

In [251]:
new_brands.isnull().sum()

barcode           0
category        155
name              0
_id.$oid          0
cpg.$id.$oid      0
cpg.$ref          0
brandCode       234
dtype: int64

In [244]:
new_brands['category'].unique()

array(['Baking', 'Beverages', 'Candy & Sweets', 'Condiments & Sauces',
       'Canned Goods & Soups', nan, 'Magazines', 'Breakfast & Cereal',
       'Beer Wine Spirits', 'Health & Wellness', 'Beauty', 'Baby',
       'Frozen', 'Grocery', 'Snacks', 'Household', 'Personal Care',
       'Dairy', 'Cleaning & Home Improvement', 'Deli',
       'Beauty & Personal Care', 'Bread & Bakery', 'Outdoor',
       'Dairy & Refrigerated'], dtype=object)

In [261]:
new_brands.shape

(1167, 7)

In [247]:
df_brands

barcode            category      categoryCode  \
0     511111019862              Baking            BAKING   
1     511111519928           Beverages         BEVERAGES   
2     511111819905              Baking            BAKING   
3     511111519874              Baking            BAKING   
4     511111319917      Candy & Sweets  CANDY_AND_SWEETS   
...            ...                 ...               ...   
1162  511111116752              Baking            BAKING   
1163  511111706328  Breakfast & Cereal               NaN   
1164  511111416173      Candy & Sweets  CANDY_AND_SWEETS   
1165  511111400608             Grocery               NaN   
1166  511111019930              Baking            BAKING   

                           name topBrand                  _id.$oid  \
0     test brand @1612366101024    False  601ac115be37ce2ead437551   
1                     Starbucks    False  601c5460be37ce2ead43755f   
2     test brand @1612366146176    False  601ac142be37ce2ead43755d   
3     test brand @1612366146051    False  601ac142be37ce2ead43755a   
4     test brand @1612366146827    False  601ac142be37ce2ead43755e   
...                         ...      ...                       ...   
1162  test brand @1601644365844      NaN  5f77274dbe37ce6b592e90c0   
1163        Dippin Dots® Cereal      NaN  5dc1fca91dda2c0ad7da64ae   
1164  test brand @1598639215217      NaN  5f494c6e04db711dd8fe87e7   
1165          LIPTON TEA Leaves    False  5a021611e4b00efe02b02a57   
1166  test brand @1613158231643    False  6026d757be37ce6369301468   

                  cpg.$id.$oid cpg.$ref                      brandCode  
0     601ac114be37ce2ead437550     Cogs                            NaN  
1     5332f5fbe4b03c9a25efd0ba     Cogs                      STARBUCKS  
2     601ac142be37ce2ead437559     Cogs  TEST BRANDCODE @1612366146176  
3     601ac142be37ce2ead437559     Cogs  TEST BRANDCODE @1612366146051  
4     5332fa12e4b03c9a25efd1e7     Cogs  TEST BRANDCODE @1612366146827  
...                        ...      ...                            ...  
1162  5f77274dbe37ce6b592e90bf     Cogs                            NaN  
1163  53e10d6368abd3c7065097cc     Cogs             DIPPIN DOTS CEREAL  
1164  5332fa12e4b03c9a25efd1e7     Cogs  TEST BRANDCODE @1598639215217  
1165  5332f5f6e4b03c9a25efd0b4     Cogs              LIPTON TEA Leaves  
1166  6026d757be37ce6369301467     Cogs  TEST BRANDCODE @1613158231644  

[1167 rows x 9 columns]

In [262]:
new_brands.isnull().sum()

barcode           0
category        155
name              0
_id.$oid          0
cpg.$id.$oid      0
cpg.$ref          0
brandCode       234
dtype: int64

In [278]:
new_brands['brandCode'].unique()

array([nan, 'STARBUCKS', 'TEST BRANDCODE @1612366146176',
       'TEST BRANDCODE @1612366146051', 'TEST BRANDCODE @1612366146827',
       'TEST BRANDCODE @1612366146091', 'TEST BRANDCODE @1612366146133',
       'J.L. KRAFT', 'CAMPBELLS HOME STYLE', 'TEST',
       'TEST BRANDCODE @1598813526777', 'CALUMET', '511111205012',
       'AUNT JEMIMA SYRUP', 'MOLSON', 'LOTRIMIN',
       'TEST BRANDCODE @1597342520277', 'ST IVES', 'CHRISIMAGE',
       'ALKA SELTZER', "JACK DANIEL'S BARBECUE", 'MAGNUM Ice Cream',
       '511111105329', 'TEST BRANDCODE @1598635634882', 'TACO BELL',
       'FROSTED CHEERIOS', 'TEST BRANDCODE @1598639199674',
       'GODIVA DRY PACKAGED DESSERTS', 'LARABAR',
       'TEST BRANDCODE @1597350074333', 'TEST BRANDCODE @1607636368717',
       'TEST BRANDCODE @1607707830095', 'COTTONELLE', 'IZZE', 'MIO',
       '511111505365', 'QUILTING SPECIAL EDITION',
       'TEST BRANDCODE @1604437351617', 'HERMAN', 'KEVITA', 'DELIMEX',
       'THE RIGHT TO SHOWER', 'CARESS', 'TEST BRA

In [288]:
new_brands['brandCode'].describe()

count     933
unique    897
top          
freq       35
Name: brandCode, dtype: object

###### Brandcode needs a lot of remodeling, as most of the data stored is not accurate and they are stored in different formats.

In [274]:
new_brands['category'].isnull().sum()

155

###### Let's see how the dataframe is going to change if we remove the rows with null values, as there are 155 rows in category which are null and 234 in brandcode.

In [284]:
new_brands_mod.isnull().sum()

barcode         0
category        0
categoryCode    0
name            0
topBrand        0
_id.$oid        0
cpg.$id.$oid    0
cpg.$ref        0
brandCode       0
dtype: int64

In [286]:
new_brands['category'].groupby(new_brands['category']).count()

category
Baby                            18
Baking                         369
Beauty                           9
Beauty & Personal Care           6
Beer Wine Spirits               90
Beverages                       63
Bread & Bakery                   5
Breakfast & Cereal              40
Candy & Sweets                  71
Canned Goods & Soups            12
Cleaning & Home Improvement      6
Condiments & Sauces             27
Dairy                           33
Dairy & Refrigerated             5
Deli                             6
Frozen                          24
Grocery                         39
Health & Wellness               44
Household                        5
Magazines                       44
Outdoor                          1
Personal Care                   20
Snacks                          75
Name: category, dtype: int64

###### There are more number of bakery items which are sold based on the current dataset.

#### Removing the rows with Null Values is not a good option as the number of categories are reducing. It leads to wrong assumptions

In [287]:
new_brands_mod['category'].groupby(new_brands_mod['category']).count()

category
Baby                             4
Baking                         150
Beer Wine Spirits               31
Beverages                        1
Bread & Bakery                   5
Candy & Sweets                  37
Cleaning & Home Improvement      1
Dairy & Refrigerated             1
Frozen                           1
Grocery                         11
Health & Wellness                9
Personal Care                    4
Name: category, dtype: int64

In [298]:
new_brands['category'].isnull().sum()

0

In [253]:
df_receipts

bonusPointsEarned                            bonusPointsEarnedReason  \
0                 500.0  Receipt number 2 completed, bonus point schedu...   
1                 150.0  Receipt number 5 completed, bonus point schedu...   
2                   5.0                         All-receipts receipt bonus   
3                   5.0                         All-receipts receipt bonus   
4                   5.0                         All-receipts receipt bonus   
...                 ...                                                ...   
1114               25.0                        COMPLETE_NONPARTNER_RECEIPT   
1115                NaN                                                NaN   
1116                NaN                                                NaN   
1117               25.0                        COMPLETE_NONPARTNER_RECEIPT   
1118                NaN                                                NaN   

     pointsEarned  purchasedItemCount  \
0           500.0                 5.0   
1           150.0                 2.0   
2               5                 1.0   
3             5.0                 4.0   
4             5.0                 2.0   
...           ...                 ...   
1114         25.0                 2.0   
1115          NaN                 NaN   
1116          NaN                 NaN   
1117         25.0                 2.0   
1118          NaN                 NaN   

                                 rewardsReceiptItemList rewardsReceiptStatus  \
0     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
1     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
2     [{'needsFetchReview': False, 'partnerItemId': ...             REJECTED   
3     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
4     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
...                                                 ...                  ...   
1114  [{'barcode': 'B076FJ92M4', 'description': 'mue...             REJECTED   
1115                                                NaN            SUBMITTED   
1116                                                NaN            SUBMITTED   
1117  [{'barcode': 'B076FJ92M4', 'description': 'mue...             REJECTED   
1118                                                NaN            SUBMITTED   

     totalSpent                    userId                  _id.$oid  \
0         26.00  5ff1e1eacfcf6c399c274ae6  5ff1e1eb0a720f0523000575   
1         11.00  5ff1e194b6a9d73a3a9f1052  5ff1e1bb0a720f052300056b   
2         10.00  5ff1e1f1cfcf6c399c274b0b  5ff1e1f10a720f052300057a   
3         28.00  5ff1e1eacfcf6c399c274ae6  5ff1e1ee0a7214ada100056f   
4          1.00  5ff1e194b6a9d73a3a9f1052  5ff1e1d20a7214ada1000561   
...         ...                       ...                       ...   
1114      34.96  5fc961c3b8cfca11a077dd33  603cc0630a720fde100003e6   
1115        NaN  5fc961c3b8cfca11a077dd33  603d0b710a720fde1000042a   
1116        NaN  5fc961c3b8cfca11a077dd33  603cf5290a720fde10000413   
1117      34.96  5fc961c3b8cfca11a077dd33  603ce7100a7217c72c000405   
1118        NaN  5fc961c3b8cfca11a077dd33  603c4fea0a7217c72c000389   

      createDate.$date  dateScanned.$date  finishedDate.$date  \
0        1609687531000      1609687531000        1.609688e+12   
1        1609687483000      1609687483000        1.609687e+12   
2        1609687537000      1609687537000                 NaN   
3        1609687534000      1609687534000        1.609688e+12   
4        1609687506000      1609687506000        1.609688e+12   
...                ...                ...                 ...   
1114     1614594147000      1614594147000                 NaN   
1115     1614613361873      1614613361873                 NaN   
1116     1614607657664      1614607657664                 NaN   
1117     1614604048000      1614604048000                 NaN   
1118     1614565354962      1614565354962      

In [306]:
# Set the limit
# Drop columns using that limit
limit = len(df_receipts) * 0.7
new_receipts = df_receipts.dropna(axis=1, thresh=limit)

In [309]:
df_receipts

bonusPointsEarned                            bonusPointsEarnedReason  \
0                 500.0  Receipt number 2 completed, bonus point schedu...   
1                 150.0  Receipt number 5 completed, bonus point schedu...   
2                   5.0                         All-receipts receipt bonus   
3                   5.0                         All-receipts receipt bonus   
4                   5.0                         All-receipts receipt bonus   
...                 ...                                                ...   
1114               25.0                        COMPLETE_NONPARTNER_RECEIPT   
1115                NaN                                                NaN   
1116                NaN                                                NaN   
1117               25.0                        COMPLETE_NONPARTNER_RECEIPT   
1118                NaN                                                NaN   

     pointsEarned  purchasedItemCount  \
0           500.0                 5.0   
1           150.0                 2.0   
2               5                 1.0   
3             5.0                 4.0   
4             5.0                 2.0   
...           ...                 ...   
1114         25.0                 2.0   
1115          NaN                 NaN   
1116          NaN                 NaN   
1117         25.0                 2.0   
1118          NaN                 NaN   

                                 rewardsReceiptItemList rewardsReceiptStatus  \
0     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
1     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
2     [{'needsFetchReview': False, 'partnerItemId': ...             REJECTED   
3     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
4     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
...                                                 ...                  ...   
1114  [{'barcode': 'B076FJ92M4', 'description': 'mue...             REJECTED   
1115                                                NaN            SUBMITTED   
1116                                                NaN            SUBMITTED   
1117  [{'barcode': 'B076FJ92M4', 'description': 'mue...             REJECTED   
1118                                                NaN            SUBMITTED   

     totalSpent                    userId                  _id.$oid  \
0         26.00  5ff1e1eacfcf6c399c274ae6  5ff1e1eb0a720f0523000575   
1         11.00  5ff1e194b6a9d73a3a9f1052  5ff1e1bb0a720f052300056b   
2         10.00  5ff1e1f1cfcf6c399c274b0b  5ff1e1f10a720f052300057a   
3         28.00  5ff1e1eacfcf6c399c274ae6  5ff1e1ee0a7214ada100056f   
4          1.00  5ff1e194b6a9d73a3a9f1052  5ff1e1d20a7214ada1000561   
...         ...                       ...                       ...   
1114      34.96  5fc961c3b8cfca11a077dd33  603cc0630a720fde100003e6   
1115        NaN  5fc961c3b8cfca11a077dd33  603d0b710a720fde1000042a   
1116        NaN  5fc961c3b8cfca11a077dd33  603cf5290a720fde10000413   
1117      34.96  5fc961c3b8cfca11a077dd33  603ce7100a7217c72c000405   
1118        NaN  5fc961c3b8cfca11a077dd33  603c4fea0a7217c72c000389   

      createDate.$date  dateScanned.$date  finishedDate.$date  \
0        1609687531000      1609687531000        1.609688e+12   
1        1609687483000      1609687483000        1.609687e+12   
2        1609687537000      1609687537000                 NaN   
3        1609687534000      1609687534000        1.609688e+12   
4        1609687506000      1609687506000        1.609688e+12   
...                ...                ...                 ...   
1114     1614594147000      1614594147000                 NaN   
1115     1614613361873      1614613361873                 NaN   
1116     1614607657664      1614607657664                 NaN   
1117     1614604048000      1614604048000                 NaN   
1118     1614565354962      1614565354962      

In [308]:
df_receipts

bonusPointsEarned                            bonusPointsEarnedReason  \
0                 500.0  Receipt number 2 completed, bonus point schedu...   
1                 150.0  Receipt number 5 completed, bonus point schedu...   
2                   5.0                         All-receipts receipt bonus   
3                   5.0                         All-receipts receipt bonus   
4                   5.0                         All-receipts receipt bonus   
...                 ...                                                ...   
1114               25.0                        COMPLETE_NONPARTNER_RECEIPT   
1115                NaN                                                NaN   
1116                NaN                                                NaN   
1117               25.0                        COMPLETE_NONPARTNER_RECEIPT   
1118                NaN                                                NaN   

     pointsEarned  purchasedItemCount  \
0           500.0                 5.0   
1           150.0                 2.0   
2               5                 1.0   
3             5.0                 4.0   
4             5.0                 2.0   
...           ...                 ...   
1114         25.0                 2.0   
1115          NaN                 NaN   
1116          NaN                 NaN   
1117         25.0                 2.0   
1118          NaN                 NaN   

                                 rewardsReceiptItemList rewardsReceiptStatus  \
0     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
1     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
2     [{'needsFetchReview': False, 'partnerItemId': ...             REJECTED   
3     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
4     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
...                                                 ...                  ...   
1114  [{'barcode': 'B076FJ92M4', 'description': 'mue...             REJECTED   
1115                                                NaN            SUBMITTED   
1116                                                NaN            SUBMITTED   
1117  [{'barcode': 'B076FJ92M4', 'description': 'mue...             REJECTED   
1118                                                NaN            SUBMITTED   

     totalSpent                    userId                  _id.$oid  \
0         26.00  5ff1e1eacfcf6c399c274ae6  5ff1e1eb0a720f0523000575   
1         11.00  5ff1e194b6a9d73a3a9f1052  5ff1e1bb0a720f052300056b   
2         10.00  5ff1e1f1cfcf6c399c274b0b  5ff1e1f10a720f052300057a   
3         28.00  5ff1e1eacfcf6c399c274ae6  5ff1e1ee0a7214ada100056f   
4          1.00  5ff1e194b6a9d73a3a9f1052  5ff1e1d20a7214ada1000561   
...         ...                       ...                       ...   
1114      34.96  5fc961c3b8cfca11a077dd33  603cc0630a720fde100003e6   
1115        NaN  5fc961c3b8cfca11a077dd33  603d0b710a720fde1000042a   
1116        NaN  5fc961c3b8cfca11a077dd33  603cf5290a720fde10000413   
1117      34.96  5fc961c3b8cfca11a077dd33  603ce7100a7217c72c000405   
1118        NaN  5fc961c3b8cfca11a077dd33  603c4fea0a7217c72c000389   

      createDate.$date  dateScanned.$date  finishedDate.$date  \
0        1609687531000      1609687531000        1.609688e+12   
1        1609687483000      1609687483000        1.609687e+12   
2        1609687537000      1609687537000                 NaN   
3        1609687534000      1609687534000        1.609688e+12   
4        1609687506000      1609687506000        1.609688e+12   
...                ...                ...                 ...   
1114     1614594147000      1614594147000                 NaN   
1115     1614613361873      1614613361873                 NaN   
1116     1614607657664      1614607657664                 NaN   
1117     1614604048000      1614604048000                 NaN   
1118     1614565354962      1614565354962      

In [307]:
new_receipts

rewardsReceiptStatus                    userId                  _id.$oid  \
0                FINISHED  5ff1e1eacfcf6c399c274ae6  5ff1e1eb0a720f0523000575   
1                FINISHED  5ff1e194b6a9d73a3a9f1052  5ff1e1bb0a720f052300056b   
2                REJECTED  5ff1e1f1cfcf6c399c274b0b  5ff1e1f10a720f052300057a   
3                FINISHED  5ff1e1eacfcf6c399c274ae6  5ff1e1ee0a7214ada100056f   
4                FINISHED  5ff1e194b6a9d73a3a9f1052  5ff1e1d20a7214ada1000561   
...                   ...                       ...                       ...   
1114             REJECTED  5fc961c3b8cfca11a077dd33  603cc0630a720fde100003e6   
1115            SUBMITTED  5fc961c3b8cfca11a077dd33  603d0b710a720fde1000042a   
1116            SUBMITTED  5fc961c3b8cfca11a077dd33  603cf5290a720fde10000413   
1117             REJECTED  5fc961c3b8cfca11a077dd33  603ce7100a7217c72c000405   
1118            SUBMITTED  5fc961c3b8cfca11a077dd33  603c4fea0a7217c72c000389   

      createDate.$date  dateScanned.$date  modifyDate.$date  
0        1609687531000      1609687531000     1609687536000  
1        1609687483000      1609687483000     1609687488000  
2        1609687537000      1609687537000     1609687542000  
3        1609687534000      1609687534000     1609687539000  
4        1609687506000      1609687506000     1609687511000  
...                ...                ...               ...  
1114     1614594147000      1614594147000     1614594148000  
1115     1614613361873      1614613361873     1614613361873  
1116     1614607657664      1614607657664     1614607657664  
1117     1614604048000      1614604048000     1614604049000  
1118     1614565354962      1614565354962     1614565354962  

[1119 rows x 6 columns]

###### By removing the columns which contain more than 70% of missing data, the number of columns reduced from 15 to 6.

In [311]:
new_receipts.isnull().sum()

rewardsReceiptStatus    0
userId                  0
_id.$oid                0
createDate.$date        0
dateScanned.$date       0
modifyDate.$date        0
dtype: int64

In [310]:
new_receipts['rewardsReceiptStatus'].value_counts()

FINISHED     518
SUBMITTED    434
REJECTED      71
PENDING       50
FLAGGED       46
Name: rewardsReceiptStatus, dtype: int64

In [269]:
# Querying can also be done in PSSQL within Jupyter Notebook.
# SQL Queries using PSSQL
q1 = """
select * from df_users where role='fetch-staff';
"""
ps.sqldf(q1, locals()).head()

active         role signUpSource state                        id  \
0       1  fetch-staff         None    WI  59c124bae4b0299e55b0f330   
1       1  fetch-staff         None    WI  59c124bae4b0299e55b0f330   
2       1  fetch-staff         None    WI  59c124bae4b0299e55b0f330   
3       1  fetch-staff         None    WI  59c124bae4b0299e55b0f330   
4       1  fetch-staff         None    WI  59c124bae4b0299e55b0f330   

  createdDate   lastLogin  
0  1970-01-01  1970-01-01  
1  1970-01-01  1970-01-01  
2  1970-01-01  1970-01-01  
3  1970-01-01  1970-01-01  
4  1970-01-01  1970-01-01

In [313]:
q1 = """
select * from df_brands where category='Baking';
"""
ps.sqldf(q1, locals()).head()

barcode category categoryCode                       name  topBrand  \
0  511111019862   Baking       BAKING  test brand @1612366101024       0.0   
1  511111819905   Baking       BAKING  test brand @1612366146176       0.0   
2  511111519874   Baking       BAKING  test brand @1612366146051       0.0   
3  511111719885   Baking       BAKING  test brand @1612366146091       0.0   
4  511111219897   Baking       BAKING  test brand @1612366146133       0.0   

                   _id.$oid              cpg.$id.$oid cpg.$ref  \
0  601ac115be37ce2ead437551  601ac114be37ce2ead437550     Cogs   
1  601ac142be37ce2ead43755d  601ac142be37ce2ead437559     Cogs   
2  601ac142be37ce2ead43755a  601ac142be37ce2ead437559     Cogs   
3  601ac142be37ce2ead43755b  601ac142be37ce2ead437559     Cogs   
4  601ac142be37ce2ead43755c  601ac142be37ce2ead437559     Cogs   

                       brandCode  
0                           None  
1  TEST BRANDCODE @1612366146176  
2  TEST BRANDCODE @1612366146051  
3  TEST BRANDCODE @1612366146091  
4  TEST BRANDCODE @1612366146133

In [316]:
q1 = """
select * from new_brands where brandCode='STARBUCKS';
"""
ps.sqldf(q1, locals()).head()

barcode   category       name                  _id.$oid  \
0  511111519928  Beverages  Starbucks  601c5460be37ce2ead43755f   

               cpg.$id.$oid cpg.$ref  brandCode  
0  5332f5fbe4b03c9a25efd0ba     Cogs  STARBUCKS

In [321]:
q1 = """
select * from new_brands where barcode='511111219897';
"""
ps.sqldf(q1, locals()).head()

barcode category                       name                  _id.$oid  \
0  511111219897   Baking  test brand @1612366146133  601ac142be37ce2ead43755c   

               cpg.$id.$oid cpg.$ref                      brandCode  
0  601ac142be37ce2ead437559     Cogs  TEST BRANDCODE @1612366146133

In [326]:
q1 = """
select count(*) from new_receipts where rewardsReceiptStatus='FINISHED';
"""
ps.sqldf(q1, locals()).head()

count(*)
0       518

# Summary:

#### There are three datasets which are users, brands, and receipts. As there are lot of null values, the data has to be either treated by removing unwanted columns in the datasets and keeping only the required columns for our further analysis based on the objective of using the dataset like improving the customers using Fetch Rewards, or performing the analysis for a specific task by performing feature selection and keeping only the required variables. I have also observed that there are lot of duplicate and null values, which might affect the analysis.

#### And I have encountered that there is no direct relationship between the item and the brand, and so created an intermediate table that maps them.

#### The columns in the tables are not names which can be readable and so it needs to be changed. The primary key and foriegn key reference to all the tables is not exact. Primary and Foreign key reference to tables should be exact and precise. The data storage procedures can be improved.

#### Creating a new table out of nested json might have errors such as null error or floating point error. And, Querying this data is hard, and does not provide informative results.


#### Some of the issues which can be resolved are by understanding about the data generating process, attributes we store and their relation. 

#### I also have some concerns with performance and scalability, when the data increases in huge amounts, we might move to a distributed environment like spark, where we can perform processing of data quickly and easily.
#### We can create an ETL pipeline if we control the entire pipeline right from generating the data to consuming it and improving the storage procedures.